### PR TITLE
feat: 855 development of seedlot review step 3 area of use

### LINKED ISSUE #855: Development of Seedlot Review Step 3 - Area of Use
Describe the task:

Create FE of Seedlot Review, Step 3 (Area of Use) and integrate endpoints as you go along.

Figma design here: https://www.figma.com/file/ROgZJQezb1DDG4AvEgjtpw/Registration-Process----Reviewer?type=design&node-id=2744-284868&mode=design&t=XxxbormmPMPHmENb-4

Image

Acceptance Criteria

- [ ] Developed as per Figma
- [ ]  relevant documentation updated
- [ ] tested and merged
- [ ] Develop floater save button place holder - design to be tested by TSC staff

### PR DESCRIPTION
**Pull request by craigyu**
# Description
Closes #855

Resolved ~KNOWN ISSUE~:
~~The delete button in edit mode is giving off errors in console, but it's because of this:
https://github.com/carbon-design-system/carbon/pull/11893
Will talk to designers about it~~

Solution: use normal tertiary icon only button but make it red

#### New
- Area of Use read and edit function

#### Changed
- A-Class context to store the users' input data for this section, this section does not exist in the registration forms so it have to be created. the context file for a-class is becoming too large, might need to find a way to streamline it, but it works for now
- The Review Seedlot button can only be seen with seedlot that aren't in "PND" or "INC" status
- Refactored GET seedlot by seedlot number endpoint so it returns additional values rather than just the row in the `seedlot` table
- `CodeDescriptionDto` is now a class instead of record so it can be extended by other dtos

#### Removed
- None

### How was this tested?
- [ ] 🧠 Not needed
- [x] 👀 Eyeball
- [ ] 🤖 Added tests

###  What gif/image best describes this PR or how it makes you feel?

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://nr-spar-49-frontend.apps.silver.devops.gov.bc.ca/)
- [Backend](https://nr-spar-1199-backend.apps.silver.devops.gov.bc.ca/swagger-ui/index.html)
- [Oracle-API](https://nr-spar-1199-oracle-api.apps.silver.devops.gov.bc.ca/actuator/health)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/nr-spar/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge.yml)

### CODE REVIEW

**Review by RMCampos — APPROVED** [review]
it looks great and works nicely! Good job 🥇 🚀

**Comment by RMCampos** [comment]
BTW, It's so nice to see all those TS types and null comparisions. Make Javascript looks so professional 😆

**Comment by craigyu** [comment]
> BTW, It's so nice to see all those TS types and null comparisions. Make Javascript looks so professional 😆

Thank you! without them it works still but it's easier to do dev work this way

**Comment by craigyu** [comment]
forgot to fix the carbon button issue from yesterday

File: backend/src/main/java/ca/bc/gov/backendstartapi/dto/SeedlotDto.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class SeedlotDto {
+
+  @Schema(description = "The seedlot entity in the seedlot table")
+  private Seedlot seedlot;
+
+  @Schema(description = "The pirmary seed plan zone")
+  private SeedPlanZoneDto priamrySpz;
```
Comment (mgaseta) [inline backend/src/main/java/ca/bc/gov/backendstartapi/dto/SeedlotDto.java:22]: Typo in `primary` (annotation and variable name).

File: frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+          Seed Planning Zone(s) (SPZ)
+        </Column>
+      </Row>
+      <Row>
+        <Column className="info-col" sm={4} md={8} lg={8}>
+          <ReadOnlyInput
+            id="primary-spz-input-readonly"
+            label="Primary seed planning zone"
+            value={
+              richSeedlotData?.priamrySpz
```
Comment (mgaseta) [inline frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx:31]: Fix field name :)

File: frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+            showSkeleton={isFetchingData}
+          />
+        </Column>
+      </Row>
+
+      <Divider />
+
+      <Row>
+        <Column className="sub-section-title-col">
+          Elevation, Latitude and longitude
```
Comment (mgaseta) [inline frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx:58]: Shouldn't `longitude` be capitalized?

File: frontend/src/types/SpzDtoType.ts [hunk 4]
```diff
@@ -0,0 +1 @@
+
```
Comment (mgaseta) [inline frontend/src/types/SpzDtoType.ts:1]: Remove empty file.

File: frontend/src/views/Seedlot/ContextContainerClassA/utils.ts [hunk 5]
```diff
@@ -914,3 +943,76 @@ export const resDataToState = (
...
+  }
+  if (seedlotData.seedlot.longitudeMinMax !== null) {
+    clonedAreaOfUse.maxLongMinute.value = seedlotData.seedlot.longitudeMinMax.toString();
+  }
+  if (seedlotData.seedlot.longitudeSecMin !== null) {
+    clonedAreaOfUse.minLongSec.value = seedlotData.seedlot.longitudeSecMin.toString();
+  }
+  if (seedlotData.seedlot.longitudeSecMax !== null) {
+    clonedAreaOfUse.maxLongSec.value = seedlotData.seedlot.longitudeSecMax.toString();
+  }
```
Comment (mgaseta) [inline frontend/src/views/Seedlot/ContextContainerClassA/utils.ts:1015]: Should we create an object or array with all fields names and transform this bunch of ifs in a for? Or something similar, make it more elegant :monocle_face:

**Review by mgaseta — CHANGES_REQUESTED** [review]
Nice job! :1st_place_medal: :100: 
Added a few markups, but nothing major :)

File: frontend/src/views/Seedlot/ContextContainerClassA/utils.ts (hunk 5 above)
Comment (craigyu) [reply frontend/src/views/Seedlot/ContextContainerClassA/utils.ts:1015]: Good idea but I think this way is easier to maintain personally, i don't like the idea that we have to maintain a list of properties to traverse to,  I also changed the some properties name so they are less confusing such as `latitudeMinMin` -> `minLatMinute`, in this case we'll have to maintain a map object instead. Regardless, I'll try to see what this solution looks like

File: frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx (hunk 3 above)
Comment (craigyu) [reply frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx:58]: Yea, bad figma, will fix
image

File: backend/src/main/java/ca/bc/gov/backendstartapi/dto/SeedlotDto.java (hunk 1 above)
Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/dto/SeedlotDto.java:22]: funny how a typo got propagated to the frontend too

File: frontend/src/views/Seedlot/ContextContainerClassA/utils.ts (hunk 5 above)
Comment (craigyu) [reply frontend/src/views/Seedlot/ContextContainerClassA/utils.ts:1015]: It's too much TypeScript gymnastic for me to spend more time on this, I'm going to keep the`if` blocks

Solution 1:
could work, but typescript complains
image
image

Solution 2:
Could work, but typescript complains
image
image

Solution 3:
Could work, but typescript COMPLAINS
image
image

Comment (mgaseta) [reply frontend/src/views/Seedlot/ContextContainerClassA/utils.ts:1015]: OK, let's leave the ifs, thanks for the effort!

File: backend/src/main/java/ca/bc/gov/backendstartapi/dto/SeedlotDto.java (hunk 1 above)
Comment (craigyu) [reply backend/src/main/java/ca/bc/gov/backendstartapi/dto/SeedlotDto.java]: https://github.com/bcgov/nr-spar/pull/1199/commits/c6a911d6360663dec0af157871ad88033bc2d2d2

File: frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx (hunk 2 above)
Comment (craigyu) [reply frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx]: https://github.com/bcgov/nr-spar/pull/1199/commits/c6a911d6360663dec0af157871ad88033bc2d2d2

File: frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx (hunk 3 above)
Comment (craigyu) [reply frontend/src/components/SeedlotReviewSteps/AreaOfUse/Read/index.tsx]: https://github.com/bcgov/nr-spar/pull/1199/commits/c6a911d6360663dec0af157871ad88033bc2d2d2

File: frontend/src/types/SpzDtoType.ts (hunk 4 above)
Comment (craigyu) [reply frontend/src/types/SpzDtoType.ts:1]: https://github.com/bcgov/nr-spar/pull/1199/commits/c6a911d6360663dec0af157871ad88033bc2d2d2

**Comment by craigyu** [comment]
@mgaseta 
sorry I didn't make separate commits for your requested changes, they all in here https://github.com/bcgov/nr-spar/pull/1199/commits/c6a911d6360663dec0af157871ad88033bc2d2d2

File: frontend/src/components/SeedlotReviewSteps/AreaOfUse/utils.ts [hunk 6]
```diff
@@ -51,3 +65,86 @@ export const filterSpzListItem = (
...
+
+const isElevationOutOfRange = (value: string) => (
+  !validator.isInt(value, { min: MIN_ELEVATION_LIMIT, max: MAX_ELEVATION_LIMIT })
+);
+
+const isMinMaxInvalid = (minVal: string, maxVal: string) => (
+  !validator.isInt(minVal, { min: Number(minVal), max: Number(maxVal) })
+);
+
+export const validateElevatioRange = (inputObj: StringInputType): StringInputType => {
```
Comment (mgaseta) [inline frontend/src/components/SeedlotReviewSteps/AreaOfUse/utils.ts:77]: Typo on `validateElevationRange` (Missing n on elevation)

File: frontend/src/components/SeedlotReviewSteps/AreaOfUse/AddtionalSpzItem.tsx [hunk 7]
```diff
@@ -36,8 +36,9 @@ const AdditionalSpzItem = (
                 invalidText="Required"
               />
               <Button
+                className="danger-tertiary-btn"
```
Comment (mgaseta) [inline frontend/src/components/SeedlotReviewSteps/AreaOfUse/AddtionalSpzItem.tsx:39]: Nice!

**Review by mgaseta — APPROVED** [review]
Nice job! :100: :rocket: :robot:

File: frontend/src/components/SeedlotReviewSteps/AreaOfUse/utils.ts (hunk 6 above)
Comment (craigyu) [reply frontend/src/components/SeedlotReviewSteps/AreaOfUse/utils.ts]: fixed!

File: backend/src/main/java/ca/bc/gov/backendstartapi/service/SeedlotService.java [hunk 8]
```diff
@@ -225,14 +227,56 @@ public Optional<Page<Seedlot>> getUserSeedlots(String userId, int pageNumber, in
...
+      }
+
+      additionalSpzList =
+          spzList.stream()
+              .filter(spz -> !spz.getIsPrimary())
+              .map(
+                  spz ->
+                      new SeedPlanZoneDto(
+                          spz.getSpzCode(), spz.getSpzDescription(), spz.getIsPrimary()))
+              .toList();
```
[the remark this answers is not in the document]
Comment (DerekRoberts) [reply backend/src/main/java/ca/bc/gov/backendstartapi/service/SeedlotService.java:286]: This isn't right.  GitHub has no default vomit emoji!!  I need to use a comment!  :vomiting_face:

**Pull request merged**